### ISSUE TITLE
Hang/Deadlock when moving effects

### ISSUE THREAD

**Issue by jkriegshauser** [xLightsSequencer/xLights#4203]
**Describe the bug**
A hang can occur sometimes when moving effects on the timeline

**To Reproduce**
(and/or link to a short video showing the problem with audio track describing what you are doing)
1. With any sequence open, select a group of effects on the timeline
2. Repeatedly hit up/down to shift them up or down to a different group of models
3. Sometimes a hang will occur

**Expected behavior**
The hang should not occur

**Screenshots**
(Note the pink selected elements)
![231207 hang screenshot](https://github.com/smeighan/xLights/assets/1697375/12cd64bb-3b1c-48a6-ab60-49b8b3fa688a)

**Versions (please complete the following information):**
 - OS: Windows 11
 - xLights version 2023.20

**Additional context**

The main thread callstack is:
```
ntdll.dll!NtWaitForAlertByThreadId()
ntdll.dll!RtlAcquireSRWLockExclusive()
msvcp140.dll!mtx_do_lock(_Mtx_internal_imp_t * mtx, const xtime * target)
Xlights.exe!MainSequencer::TagAllSelectedEffects()
Xlights.exe!EffectsGrid::MoveSelectedEffectDown(bool shift)
Xlights.exe!MainSequencer::OnCharHook(wxKeyEvent & event)
Xlights.exe!wxAppConsoleBase::CallEventHandler(wxEvtHandler * handler, wxEventFunctor & functor, wxEvent & event)
Xlights.exe!wxEvtHandler::ProcessEventIfMatchesId(const wxEventTableEntryBase & entry, wxEvtHandler * handler, wxEvent & event)
Xlights.exe!wxEvtHandler::SearchDynamicEventTable(wxEvent & event)
Xlights.exe!wxEvtHandler::TryHereOnly(wxEvent & event)
[Inline Frame] Xlights.exe!wxEvtHandler::TryBeforeAndHere(wxEvent &)
[Inline Frame] Xlights.exe!wxEvtHandler::ProcessEventLocally(wxEvent &)
Xlights.exe!wxEvtHandler::ProcessEvent(wxEvent & event)
Xlights.exe!wxEvtHandler::SafelyProcessEvent(wxEvent & event)
Xlights.exe!wxKeyboardHook(int nCode, unsigned __int64 wParam, __int64 lParam)
user32.dll!DispatchHookW()
user32.dll!CallHookWithSEH(struct _GENERICHOOKHEADER *,void *,unsigned long *,unsigned __int64)
user32.dll!__fnHkINDWORD()
ntdll.dll!KiUserCallbackDispatcherContinue()
win32u.dll!NtUserPeekMessage()
user32.dll!_PeekMessage()
user32.dll!PeekMessageW()
Xlights.exe!wxMSWEventLoopBase::GetNextMessageTimeout(tagMSG * msg, unsigned long timeout)
Xlights.exe!wxMSWEventLoopBase::GetNextMessage(tagMSG * msg)
Xlights.exe!wxGUIEventLoop::Dispatch()
[Inline Frame] Xlights.exe!wxEventLoopManual::ProcessEvents()
Xlights.exe!wxEventLoopManual::DoRun()
Xlights.exe!wxEventLoopBase::Run()
Xlights.exe!wxAppConsoleBase::MainLoop()
Xlights.exe!wxEntryReal(int & argc, wchar_t * * argv)
Xlights.exe!wxEntry(int & argc, wchar_t * * argv)
[Inline Frame] Xlights.exe!invoke_main()
Xlights.exe!__scrt_common_main_seh()
kernel32.dll!BaseThreadInitThunk()
ntdll.dll!RtlUserThreadStart()
```

The mutex that is attempting to be locked is held by a thread with callstack:
```
ntdll.dll!NtWaitForAlertByThreadId()
ntdll.dll!RtlSleepConditionVariableSRW()
KERNELBASE.dll!SleepConditionVariableSRW()
[Inline Frame] msvcp140.dll!Concurrency::details::stl_condition_variable_win7::wait_for(Concurrency::details::stl_critical_section_interface *)
msvcp140.dll!Concurrency::details::stl_condition_variable_win7::wait(Concurrency::details::stl_critical_section_interface * lock)
msvcp140.dll!_Cnd_wait(_Cnd_internal_imp_t * cond, _Mtx_internal_imp_t * mtx)
Xlights.exe!xLightsFrame::RenderEffectFromMap(bool suppress, Effect * effectObj, int layer, int period, SettingsMap & SettingsMap, PixelBufferClass & buffer, bool & resetEffectState, bool bgThread, RenderEvent * event)
Xlights.exe!RenderJob::ProcessFrame(int frame, std::shared_ptr<Element> el, EffectLayerInfo & info, PixelBufferClass * buffer, int strand, bool blend)
Xlights.exe!RenderJob::Process()
Xlights.exe!JobPoolWorker::ProcessJob(Job * job)
Xlights.exe!JobPoolWorker::Entry()
Xlights.exe!startFunc(JobPoolWorker * jpw)
Xlights.exe!std::thread::_Invoke<std::tuple<void (__cdecl*)(JobPoolWorker *),JobPoolWorker *>,0,1>(void * _RawVals)
ucrtbase.dll!thread_start<unsigned int (__cdecl*)(void *),1>()
kernel32.dll!BaseThreadInitThunk()
ntdll.dll!RtlUserThreadStart()
```
This thread appears to be waiting on a condition_variable that can only be triggered by the main thread. Hence a deadlock occurs.

**Comment by AzGilrock** [comment]
Might could avoid this by turning off rendering when you need to rapidly move groups of effects up and down in fast succession.

**Comment by jkriegshauser** [comment]
If memory serves, disabling rendering does work around the hang, but that's
not a great workflow.

**Comment by AzGilrock** [comment]
Yeah sorry its a difficult problem to solve for xLights.  No other sequencer has the capabilities that are involved with that move operation.  What do I mean?  Well most people don't realize the same effect can show up several times in the save view.  I've shown a screenshot where I had the same effect in 5 places...Lets use that as an example.  That can happen because an effect can be on a model that is a member of 5 groups.  You expand those 5 groups and that effect appears 5 times.  So what happens when you select that effect?  It gets selected in every spot because its the same effect and now its status is "selected".  Then you try to move the effect up.  That causes xLights to have to check every row above all 5 locations.  In your screenshot with that many effects (at least 44) most of the effects are actually blocking the effect below it from moving up.  So there is logic to try to figure out that the selected effects are actually going to be moving and shouldn't block.  Now you add rendering to the mix and every time you move effects it wants to halt any previous rendering and restart new rendering in the new location.  Multiply that by 44.  We could just make it so nothing renders until to command it and that would get rid of a lot of these issues but then people would complain their view is never up-to-date until they hit render.  Currently you have the option.  You can disable that rendering and then hit render all to update everything.  Would like us to force it so that's how it always works?

**Comment by jkriegshauser** [comment]
@AzGilrock that's a good point, and actually a great description of the perfect use case for using `shared_ptr`, which [I proposed](https://github.com/xLightsSequencer/xLights/pull/4135) but was shut down.

I do a _lot_ of multithreaded C++ for work and would be interested in helping out. A good approach to solving problems like this is to use a shared memory model for objects where if the conditions for their lifecycle changes the thread working on them can keep working on them but then their results get discarded later. In this way you can be _eventually consistent_ with respect to the rendered effect without having to "stop the world" to make sure that it's correct right now (leading to hitches).

**Comment by AzGilrock** [comment]
I haven't looked at the code changes you suggested I just saw all the comments around it.  I do tons of multi-threaded work myself.  I'm not sure it would work out well having one thread able to change an effects settings at the same time another thread is trying to use those settings for rendering.  The whole reason we have the mutex is because multiple threads used to access the objects concurrently and caused crashing.  I'll admit I learned to code way before shared_ptr was even added to the language and it was only recently that I moved onto a program that allowed its use so I'm weak on knowing the ins and out of that part.

**Comment by jkriegshauser** [comment]
@AzGilrock [shared_ptr](https://en.cppreference.com/w/cpp/memory/shared_ptr) (and its lesser known cousin [weak_ptr](https://en.cppreference.com/w/cpp/memory/weak_ptr)) are all about reference counting, which is a long-used memory model and an underlying tenet of technologies like COM for instance from back in the 90s. The disconnect I think is this:
> I'm not sure it would work out well having one thread able to change an effects settings at the same time another thread is trying to use those settings for rendering.

This is exactly what we have now and the locking (especially in destructors) leads to deadlocks.

With reference counting (read: `shared_ptr`) the premise is that the main thread changes an effect by releasing the reference on the previous effect and creating a new effect (which happens automatically through `shared_ptr` being smart). The thread rendering the effect still uses the old version of the effect because it has a `shared_ptr` that retains a reference to it. The old effect may still be rendered for a few frames after changed, but the concept of _eventual consistency_ states that it will shortly pick up the changes and begin rendering the new effect. The only locking that is required is to look up the current version of the effect and add a reference to it.

**Comment by AzGilrock** [comment]
No that's not what we have now.  We don't allow both things to happen at the same time.  If we did there wouldn't be a deadlock happening.

**Comment by AzGilrock** [comment]
Back to the original post above....the mutex that you are showing is for TagAllSelectedEffects.  That mutex is only for preventing re-entry into that function it should not be holding off rendering.  When you goto move effects like I said earlier I have to tag them all to figure out which ones need to move and are allowed to move.

**Comment by jkriegshauser** [comment]
The lock that I mentioned in a destructor that I mentioned is `~ModelElement` and the deadlock for that was #4134, so that was a different issue.

Deadlocks happen because you have two or more wait targets (typically locks) and a cycle in the wait order. In this case, the main thread is possibly holding some lock ('A') and then waiting on a lock in `EffectLayer::TagAllSelectedEffects()` ('B').

The render thread is likely holding lock ('B') and then waiting on a `condition_variable` that is only notified from the main thread (while not really a lock, this is "lock" ('A')). The main thread is blocked and will never notify the `condition_variable`, leading to deadlock. Holding locks while waiting on a different condition is dangerous.

The `shared_ptr` discussion applies more to #4134 as it requires significantly fewer locks. In this case, it seems like the render thread should not be holding mutex ('B') while waiting for the `condition_variable`.

**Comment by AzGilrock** [comment]
I took a quick look and this issue could probably be prevented by having all the calls like MoveSelectedEffectUp do an abortRender call first.  I'm pretty sure that function is trying to tag all the effects but hasn't stopped rendering.  Way back when I first wrote this code to move effects around we weren't seeing these lockups.  Probably the effects were rendering so fast and we all had smaller shows back then.

**Comment by AzGilrock** [comment]
Can you see if you can still get it to lockup with the latest version?  I went to add some changes to the TagAllSelectedEffects and found Dan had already done work on this a couple weeks ago in this commit:
https://github.com/xLightsSequencer/xLights/commit/82493f81b3cd621d2a376a2b6cd6fcd5beeeac12

**Comment by ManiacTwister** [comment]
@AzGilrock i quickly tested the 2024.03 release and at least until now i can't reproduce my issue from #4254 which seem to be similar to this one. I suppose this has something to do with your commit? 😀

**Comment by derwin12** [comment]
Potential fix in place .. you can still swamp xlights if you have lots of shaders as they need to compile.
